src/Model/LazyBSONArray.php

```diff
@@ -0,0 +1,309 @@
+    }
+
+    /** @return ListIterator<TValue> */
+    public function getIterator(): ListIterator
+    {
+        $itemIterator = new AppendIterator();
+        // Iterate through all fields in the BSON array
+        $itemIterator->append($this->bson->getIterator());
+        // Then iterate over all fields that were set
+        $itemIterator->append(new ArrayIterator($this->set));
+
+        /** @var array<int, bool> $seen */
+        $seen = [];
+
+        // Use ListIterator to ensure we're indexing from 0 without gaps
+        return new ListIterator(
+            new CallbackIterator(
+                // Skip keys that were unset or handled in a previous iterator
+                new CallbackFilterIterator(
+                    $itemIterator,
+                    /** @param TValue $value */
+                    function ($value, int $offset) use (&$seen): bool {
+                        return ! isset($this->unset[$offset]) && ! isset($seen[$offset]);
+                    },
+                ),
+                /**
+                 * @param TValue $value
+                 * @return TValue
+                 */
+                function ($value, int $offset) use (&$seen) {
+                    // Mark key as seen, skipping any future occurrences
+                    $seen[$offset] = true;
+
+                    // Return actual value (potentially overridden by offsetSet)
+                    return $this->offsetGet($offset);
+                },
+            ),
+        );
+    }
+
+    public function jsonSerialize(): array
+    {
+        return iterator_to_array($this->getIterator());
+    }
+
+    /** @param mixed $offset */
+    public function offsetExists($offset): bool
+    {
+        if (! is_numeric($offset)) {
+            return false;
+        }
+
+        $offset = (int) $offset;
+
+        return $this->exists[$offset] ??= $this->bson->has($offset);
+    }
+
+    /**
+     * @param mixed $offset
+     * @return TValue
+     */
+    #[ReturnTypeWillChange]
+    public function offsetGet($offset)
+    {
+        if (! is_numeric($offset)) {
+            trigger_error(sprintf('Undefined offset: %s', $offset), E_USER_WARNING);
+
+            return null;
+        }
+
+        $offset = (int) $offset;
+        $this->readFromBson($offset);
+
+        if (isset($this->unset[$offset]) || ! $this->exists[$offset]) {
+            trigger_error(sprintf('Undefined offset: %d', $offset), E_USER_WARNING);
+
+            return null;
+        }
+
+        return array_key_exists($offset, $this->set) ? $this->set[$offset] : $this->read[$offset];
+    }
+
+    /**
+     * @param mixed  $offset
+     * @param TValue $value
+     */
+    public function offsetSet($offset, $value): void
+    {
+        if ($offset === null) {
+            $this->readEntirePackedArray();
+
+            $existingItems = [
+                ...array_keys($this->read),
+                ...array_keys($this->set),
+            ];
```

Comment:
Since we're just interested in the maximum key, we don't need to de-dupe it. I'll make a note to revisit this when we have performance tests, as appending to a large array may actually benefit from de-duplicating (assuming that calling `array_unique` makes `max` significantly quicker to offset the additional call).

Reply:
Feel free to resolve on your own.